File: gosdk/sso.go
package gosdk

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
)

type SSO struct {
	AppID     string `yaml:"appid"`
	SecretKey string `yaml:"secretKey"`
	API       string `yaml:"api"`
}

func (s *SSO) GetLoginPage(params map[string]string, backTo ...string) (string, error) {

	urlObj, err := url.Parse(s.API)
	if err != nil {
		return "", err
	}
	query := urlObj.Query()
	for k, v := range params {
		query.Set(k, v)
	}
	if len(backTo) > 0 {
		query.Set("__backto__", backTo[0])
	}
	urlObj.Path = "/authorize"
	urlObj.RawQuery = query.Encode()

	sign := SignSDK{
		AppID:     s.AppID,
		SecretKey: s.SecretKey,
	}
	return sign.GetSignedURL(urlObj.String())

}

func (s *SSO) GetResource(scope string, token string, data interface{}) error {

	urlObj, err := url.Parse(s.API)
	if err != nil {
		return err
	}
	urlObj.Path = "/authorize/resource"
	query := urlObj.Query()
	query.Set("scope", scope)
	query.Set("resource_token", token)
	urlObj.RawQuery = query.Encode()
	sign := SignSDK{
		AppID:     s.AppID,
		SecretKey: s.SecretKey,
	}

	targetURL, err := sign.GetSignedURL(urlObj.String())
	if err != nil {
		return err
	}

	response, err := http.Get(targetURL)
	if err != nil {
		return err
	}

	defer response.Body.Close()

	body, err := ioutil.ReadAll(response.Body)

	if err != nil {
		return err
	}
	if response.StatusCode != 200 {
		return fmt.Errorf("status code: %d message: %s", response.StatusCode, string(body))
	}

	return json.Unmarshal(body, data)
}

File: nodejs/src/sso.ts
import { URL } from "url"
import Sign from "./sign"
import * as http from "http"
import * as https from "https"

export default class SSO {
    // app id
    private appid: string;
    // secret key 密钥
    private sk: string;
    // sso 服务器域名
    private api: string;
    constructor(settings: { appid: string; secretKey: string, api?: string }) {
        this.appid = settings.appid;
        this.sk = settings.secretKey;
        this.api = settings.api || "https://sso.jx3box.com";

    }
    // 获取sso登录页地址
    public getLoginPage(params: { [key: string]: string }, backto?: string): string {
        const urlObj = new URL(this.api)
        const query = urlObj.searchParams
        Object.keys(params).forEach(key => {
            query.set(key, params[key])
        })
        query.set("__backto__", backto || "")
        urlObj.pathname = "/authorize"
        urlObj.search = query.toString()
        const sign = new Sign(this.appid, this.sk)
        return sign.getSignedURL(urlObj.toString())
    }
    // 获取资源 如用户信息等
    public getResource(scope: string, token: string): Promise<any> {
        return new Promise<any>((resolve, reject) => {
            const urlObj = new URL(this.api)
            urlObj.pathname = "/authorize/resource"
            const query = urlObj.searchParams
            query.set("scope", scope)
            query.set("resource_token", token)
            urlObj.search = query.toString()
            const sign = new Sign(this.appid, this.sk)
            const targetURL = sign.getSignedURL(urlObj.toString())
            const responseHandle = (res: http.IncomingMessage) => {
                res.setEncoding('utf8');
                let rawData = '';
                res.on('data', (chunk: string) => { rawData += chunk; });
                res.on('end', () => {
                    if (res.statusCode != 200) {
                        reject(`status code: ${res.statusCode} error message: ${rawData}`);
                        return
                    }
                    try {
                        const parsedData = JSON.parse(rawData);
                        resolve(parsedData)
                    } catch (e) {
                        reject(`parse json error: ${rawData}`);
                    }
                });
            }
            const errorHandler = (err: Error) => {
                reject(`request error: ${err.message}`);
            }
            switch (urlObj.protocol) {
                case "https:":
                    https.get(targetURL, responseHandle).on('error', errorHandler);
                    break
                case "http:":
                    http.get(targetURL, responseHandle).on('error', errorHandler);
            }
        })
    }

}